_posts/2019-06-04-loading-large-image-data-with-dask-array.md

```diff
@@ -0,0 +1,150 @@
+steps, it may be very difficult if not impossible to reuse an existing
+pipeline. In other words, there is friction between the loading and analysis
+steps, which frustrates efforts to make things reusable.
+
+Having a modular and general way to load data makes it easy to present data
+stored differently in a standard way. Further having a standard way to present
+data to analytical pipelines allows that part of the pipeline to focus on what
+it does best, analysis! In general, this should decouple these to components in
+a way that improves the experience of users involved in all parts of the
+pipeline.
+
+
+Loading image data with Dask
+----------------------------
+
+A pretty typical case to see is a large directory of image files. In particular
+TIFF files are quite common, though other formats could show up as well. It is
+common to see a naming structure that mirrors that shown below. The `<i>`
+represent different numbers.
+
+```
+mydata_ch<i>_<j>t_<k>x_<l>y_<m>z.tif
+```
+
+To load a single one of these a user might right some code like this:
+
+```python
+from skimage.io import imread
+a = imread("mydata_ch<i>_<j>t_<k>x_<l>y_<m>z.tif")
+```
+
+While this works fine for reading a single image into memory, this will quickly
+become prohibitive once we try to load more data than we have memory. So we
+might replace this with something like this.
+
+```python
+from dask.array.image import imread
+a = imread("mydata_ch<i>_<j>t_<k>x_<l>y_<m>z.tif")
+```
+
+This offloads the actual image reading to scikit-image under the hood (much
+like what happened explicitly before). Though this doesn't keep the data in
+memory. Instead we briefly load the image once to figure out the metadata
+(shape and type of the array loaded) and then drop it from memory. Once we
+actually need this data for computation of some kind, it will be loaded back
+in.
+
+Though it's possible we run into data that we want to load that scikit-image
+doesn't know to handle. In this case we have a few options. Though one pretty
+easy one is we could supply our own [`imread` function for Dask to use](
+https://docs.dask.org/en/latest/array-api.html#dask.array.image.imread ). This
+way we could manage the subtleties of loading our own data image data format.
+It might involve accessing some database to pull some image data out or could
+involve interacting with cloud storage to pull some data down.
+
+Thus far we have only showed how one might load one chunk. To load multiple
+chunks we would need to iterate over our data some how and combine the chunks
+together. Suppose we have one dimension to combine chunks along, we might write
+some code that looks like that below. We would load all of the chunks of data
+we have join them together with Dask Array's `block` to form a larger array. We
+can easily extend this to more dimensions.
+
+```python
+data = []
+for fn in files:
+    data.append(dask.array.imread(fn))
```

Comment:
Are we reading each image here to get the shape and dtype?  I would expect this to become prohibitively expensive.  

When I do this I tend to do the following:

```python
sample = skimage.io.imread(fn)

data = [dask.delayed(skimage.io.imread)(fn) for fn in filenames]
data = [da.from_array(x, shape=sample.shape, dtype=sample.dtype) for x in data]
```

Reply:
Currently, yes. We could certainly change that. Though I regularly encountered use cases before when the image files did not necessarily have the same size.

Perhaps we could say, "if we know the image files to be uniform, we could inspect one for metadata and apply this to the rest like so."

Should add this is one of the things that should be in future work. In particular having a fast way to determine the metadata without reading the full image file into memory. Maybe we should hint at that here. Under the hood dask-image's `imread` does this and its quite fast, but doesn't handle all image file types.

xref: https://github.com/imageio/imageio/issues/382
xref: https://github.com/imageio/imageio/issues/362
xref: https://github.com/imageio/imageio/issues/263

Reply:
Any follow-up here, @mrocklin?

Reply:
> Perhaps we could say, "if we know the image files to be uniform, we could inspect one for metadata and apply this to the rest like so."

Sounds good to me.

> Under the hood dask-image's imread does this and its quite fast, but doesn't handle all image file types.

Ah, I assumed thhat it would be reading all of the data.  If it's fast, then great.  

I still think that I have a slight preference for delayed just because it's a bit more generalizable and explicit, but whatever works.

> 
```suggestion
    data.append(dask.array.imread(fn))
```

Reply:
Ok, I'll give that a go.

Yeah dask-image's reader is fast, but `dask.array`s is not. We could change that in the example. I think I noted this in "Future Work" currently, but we could replace that.

Sure one thought on the progression might be...

1. Load file with `skimage.io.imread`.
2. Note this won't scale to all of our data.
3. Show example with `delayed`.
4. We can brush this under the rug with dask-image's `imread`.

Does that seem reasonable?

Side note: The suggested change doesn't seem to have anything in the diff?

Reply:
Yeah, that sounds great.   We're explicit so readers can generalize, but also point to the existence  of an automatic solution so that users know where to investigate further.

Reply:
Ok, I've actually done this above with loading a single image and just changed the code here to use `imread`.